crepecake/src/integration-test/java/com/google/cloud/tools/crepecake/registry/DockerCredentialRetrieverIntegrationTest.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.registry;
+
+import com.fasterxml.jackson.databind.exc.MismatchedInputException;
+import com.google.cloud.tools.crepecake.http.Authorization;
+import java.io.IOException;
+import org.junit.Assert;
+import org.junit.Test;
+
+/** Integration tests for {@link DockerCredentialRetriever}. */
+public class DockerCredentialRetrieverIntegrationTest {
+
+  /** Tests retrieval via {@code docker-credential-gcr} CLI. */
+  @Test
+  public void testRetrieveGCR() throws IOException {
+    try {
+      DockerCredentialRetriever dockerCredentialRetriever =
+          new DockerCredentialRetriever("gcr.io", "gcr");
+
+      Authorization authorization = dockerCredentialRetriever.retrieve();
+
+      // Checks that some token was received.
+      Assert.assertTrue(0 < authorization.getToken().length());
+
+    } catch (MismatchedInputException ex) {
+      if (!ex.getMessage().contains("No content to map due to end-of-input")) {
```

Comment:
why do you allow these failures?

Reply:
Here, I am giving a more descriptive error for test failure cases where it's not necessarily the code is invalid but rather the environment that does not have the necessary components to run the test correctly.

Reply:
That's an antipattern. It does come up occasionally but please try really, really hard not to do this. If you have to, then the test should be reported as skipped, not failed. Assume class can help with this. 

But probably we should just fail if the environment is deficient.

Reply:
Do you mean it's an antipattern to fail when the environment is deficient or to catch exceptions just for providing a different exception message?

Reply:
Failing when the environment is deficient. In general, tests should not be environment dependent. 

On occasion we have tests that only apply in one platform (e.g. Windows) but in that case if you run them on Unix they should be marked skipped, not passed.

Other conditions like bad network access should be failures.

Reply:
@loosebazooka I'll revert that change and mark as skipped then?

Reply:
In this case I don't think you've justified marking it as skipped. This looks like a straight test failure to me.

Reply:
So here, if the credential store doesn't have credentials for a certain server URL, that means that the environment wasn't set up correctly (ie the credentials for that server URL wasn't put into the credential store).